tools/querytee/query_splitter.go

```diff
@@ -0,0 +1,394 @@
+	"github.com/grafana/loki/v3/pkg/querier/queryrange/queryrangebase"
+)
+
+// isQueryEntirelyRecent checks if the entire query is for recent data (newer than the threshold).
+// If true, the query should not be sent to goldfish at all.
+func isQueryEntirelyRecent(r *http.Request, minAge time.Duration) bool {
+	if minAge == 0 || r.URL.Path != "/loki/api/v1/query_range" {
+		return false
+	}
+
+	if err := r.ParseForm(); err != nil {
+		return false
+	}
+
+	rangeQuery, err := loghttp.ParseRangeQuery(r)
+	if err != nil {
+		return false
+	}
+
+	now := time.Now()
+	threshold := now.Add(-minAge)
+
+	// If the query start is at or after the threshold, entire query is recent
+	return rangeQuery.Start.After(threshold) || rangeQuery.Start.Equal(threshold)
+}
+
+// shouldSplitQuery determines if a query_range query needs to be split based on the minimum age threshold.
+// Returns whether splitting is needed, the aligned split point, and the step duration.
+// If needsSplit is false, the query should be handled normally (either all recent or all old).
+func shouldSplitQuery(r *http.Request, minAge time.Duration) (needsSplit bool, splitPoint time.Time, step time.Duration, err error) {
+	if minAge == 0 {
+		return false, time.Time{}, 0, nil
+	}
+
+	if r.URL.Path != "/loki/api/v1/query_range" {
+		return false, time.Time{}, 0, nil
+	}
+
+	if err := r.ParseForm(); err != nil {
+		return false, time.Time{}, 0, fmt.Errorf("failed to parse form: %w", err)
+	}
+
+	rangeQuery, err := loghttp.ParseRangeQuery(r)
+	if err != nil {
+		return false, time.Time{}, 0, fmt.Errorf("failed to parse range query: %w", err)
+	}
+
+	now := time.Now()
+	threshold := now.Add(-minAge)
+
+	if rangeQuery.Start.After(threshold) || rangeQuery.Start.Equal(threshold) {
+		return false, time.Time{}, 0, nil
+	}
+
+	if rangeQuery.End.Before(threshold) || rangeQuery.End.Equal(threshold) {
+		return false, time.Time{}, 0, nil
+	}
+
+	alignedSplit := alignToStep(threshold, rangeQuery.Start, rangeQuery.Step)
+	return true, alignedSplit, rangeQuery.Step, nil
+}
+
+// alignToStep aligns a split time to the nearest step boundary at or before the split time.
+// This ensures the split point falls exactly on a timestamp where Loki would evaluate the query.
+func alignToStep(splitTime time.Time, queryStart time.Time, step time.Duration) time.Time {
+	if step <= 0 {
+		return splitTime
+	}
+
+	if splitTime.Before(queryStart) {
+		return queryStart
+	}
+
+	elapsed := splitTime.Sub(queryStart)
+	numSteps := elapsed / step
+
+	alignedSplit := queryStart.Add(numSteps * step)
+	return alignedSplit
+}
+
+// createSplitRequests creates two sub-requests from the original request:
+// - oldQuery: from start to splitPoint (inclusive)
+// - recentQuery: from (splitPoint + step) to end (no overlap)
+// Both requests preserve all other parameters from the original.
+func createSplitRequests(original *http.Request, splitPoint time.Time, step time.Duration) (oldQuery, recentQuery *http.Request, err error) {
```

Comment:
I noticed we're duplicating quite a bit of logic for stuff that already exists in other packages.

I wonder if it might make our lives easier if we wrote our logic as if we were a tripperware, such that:

1. We parse the request into a `queryrangebase.Request` 
2. We split the requests, similar to what's done in `queryrange.engineRouter` 
3. We proxy requests via `frontend.NewDownstreamRoundTripper` (which turns them into HTTP requests) 
4. We convert the response back with the codec 

I think that would still allow you to do the ultimate conversion back to BackendResponse, while also allowing us to make use of all the existing logic for this stuff. 

WDYT?

Reply:
If you think it'd be too much work to move over to tripperware for the time splitting path, I'm happy for the refactor to be a follow up